File: orko-ui/src/components/Job.js
/*
 * Orko
 * Copyright © 2018-2019 Graham Crockford
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React, { Component } from "react"
import { Icon } from "semantic-ui-react"

import * as jobTypes from "../services/jobTypes"
import JobStage from "./primitives/JobStage"
import JobStages from "./primitives/JobStages"

const BUY = "BUY"
const SELL = "SELL"

export default class Job extends Component {
  render() {
    const job = this.props.job
    if (job.jobType === jobTypes.OCO) {
      return (
        <JobStages>
          <JobStage>
            <Icon name="eye" /> Watch price{" "}
            {job.tickTrigger.base + "/" + job.tickTrigger.counter} on{" "}
            {job.tickTrigger.exchange}
            <JobStages>
              {job.high && (
                <JobStage>
                  <Icon name="pointing up" />
                  If price rises above {job.high.thresholdAsString}
                  <Job job={job.high.job} />
                </JobStage>
              )}
              {job.low && (
                <JobStage>
                  <Icon name="pointing down" />
                  If price drops below {job.low.thresholdAsString}
                  <Job job={job.low.job} />
                </JobStage>
              )}
            </JobStages>
          </JobStage>
        </JobStages>
      )
    } else if (job.jobType === jobTypes.LIMIT_ORDER) {
      return (
        <JobStages>
          <JobStage>
            <Icon name={job.direction === BUY ? "caret up" : "caret down"} />
            {job.direction} <b>{job.amount}</b> {job.tickTrigger.base} for{" "}
            {job.tickTrigger.counter} on <b>{job.tickTrigger.exchange}</b> at{" "}
            <b>{job.limitPrice}</b>
          </JobStage>
        </JobStages>
      )
    } else if (
      job.jobType === jobTypes.SOFT_TRAILING_STOP &&
      job.direction === SELL
    ) {
      return (
        <JobStages>
          <JobStage>
            <Icon name="eye" /> Watch price{" "}
            {job.tickTrigger.base + "/" + job.tickTrigger.counter} on{" "}
            {job.tickTrigger.exchange}
            <JobStages>
              <JobStage>
                <Icon name="pointing down" />
                If the bid price drops below {job.stopPrice}
                <JobStages>
                  <JobStage>
                    <Icon name="caret down" />
                    Sell at {job.limitPrice}
                  </JobStage>
                </JobStages>
              </JobStage>
              <JobStage>
                <Icon name="pointing up" />
                If the price rises
                <JobStages>
                  <JobStage>
                    <Icon name="caret up" />
                    Raise the stop price by the same amount, but not the limit
                    price. Highest bid price recorded so far is{" "}
                    {job.lastSyncPrice}
                  </JobStage>
                </JobStages>
              </JobStage>
            </JobStages>
          </JobStage>
        </JobStages>
      )
    } else if (
      job.jobType === jobTypes.SOFT_TRAILING_STOP &&
      job.direction === BUY
    ) {
      return (
        <JobStages>
          <JobStage>
            <Icon name="eye" /> Watch price{" "}
            {job.tickTrigger.base + "/" + job.tickTrigger.counter} on{" "}
            {job.tickTrigger.exchange}
            <JobStages>
              <JobStage>
                <Icon name="pointing up" />
                If ask price rises above {job.stopPrice}
                <JobStages>
                  <JobStage>
                    <Icon name="caret up" />
                    Buy at {job.limitPrice}
                  </JobStage>
                </JobStages>
              </JobStage>
              <JobStage>
                <Icon name="pointing down" />
                If the price drops
                <JobStages>
                  <JobStage>
                    <Icon name="caret down" />
                    Lower the stop price by the same amount, but not the limit
                    price. Lowest ask price recorded so far is{" "}
                    {job.lastSyncPrice}
                  </JobStage>
                </JobStages>
              </JobStage>
            </JobStages>
          </JobStage>
        </JobStages>
      )
    } else if (job.jobType === jobTypes.ALERT) {
      return (
        <JobStages>
          <JobStage>
            <Icon name="computer" />
            Send a telegram message: '{job.notification.message}'
          </JobStage>
        </JobStages>
      )
    } else if (job.jobType === jobTypes.STATUS_UPDATE) {
      return (
        <JobStages>
          <JobStage>
            <Icon name="computer" />
            Status update
          </JobStage>
        </JobStages>
      )
    } else if (job.jobType === jobTypes.SCRIPT) {
      return (
        <>
          <h3>{job.name}</h3>
          <pre>{job.script}</pre>
        </>
      )
    } else {
      return (
        <JobStages>
          <JobStage>
            <Icon name="computer" />
            {JSON.stringify(job, null, 2)}
          </JobStage>
        </JobStages>
      )
    }
  }
}

File: orko-job/src/main/java/com/gruelbox/orko/jobrun/spi/Job.java
/**
 * Orko - Copyright © 2018-2019 Graham Crockford
 *
 * <p>This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU Affero General Public License as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * <p>You should have received a copy of the GNU Affero General Public License along with this
 * program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.gruelbox.orko.jobrun.spi;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver;

/**
 * A job may start multiple times, but only run exactly once to completion. It can be serialized,
 * and thus is persistent until finishing. Its behaviour must be idempotent to avoid
 * double-processing.
 *
 * <p>Instances of this class are immutable, but can be copied and modified (using {@link
 * #toBuilder()}) and then saved, replacing the existing job.
 *
 * <p>The processing code is kept separate from the data. The applicatiom will inject an instance of
 * {@link #processorFactory()} to create the processor necessary to run the job.
 *
 * <p>In a multi instance application, only one instance can lock and run the job at a time. In the
 * event of a failure of the instance holding the lock, the lock will time out and get picked up by
 * another instance.
 *
 * <p>Jobs are assumed to be asynchronous. They should return immediately on startup and either
 * subscribe to the necessary callbacks or spawn background threads to achieve continuous
 * processing.
 *
 * @author Graham Crockford
 */
@JsonTypeInfo(
    use = JsonTypeInfo.Id.CUSTOM,
    include = JsonTypeInfo.As.PROPERTY,
    property = "jobType")
@JsonTypeIdResolver(JobTypeResolver.class)
public interface Job {

  /** @return The job id. */
  String id();

  /**
   * Creates a duplicate copy of the job as a builder, allowing a modified version to be created.
   *
   * <p>Subclasses should overload the return type of this method to match the concrete job type.
   *
   * @return A populated builder.
   */
  JobBuilder<? extends Job> toBuilder();

  /**
   * The class of the factory from which instances of the processor class can be created.
   *
   * @return The processor factory class.
   */
  Class<? extends JobProcessor.Factory<? extends Job>> processorFactory();
}
